ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,475 @@
+        cdef int ucp_epoll_fd
+        cdef epoll_event ev
+        cdef int err
+        status = ucp_worker_get_efd(self._handle, &ucp_epoll_fd)
+        assert_ucs_status(status)
+        self.arm()
+        epoll_fd = epoll_create(1)
+        assert(epoll_fd != -1)
+        ev.data.fd = ucp_epoll_fd
+        ev.data.ptr = NULL
+        ev.data.u32 = 0
+        ev.data.u64 = 0
+        ev.events = EPOLLIN
+        err = epoll_ctl(epoll_fd, EPOLL_CTL_ADD, ucp_epoll_fd, &ev)
+        assert(err == 0)
+        return epoll_fd
+
+    def arm(self):
+        cdef ucs_status_t status
+        status = ucp_worker_arm(self._handle)
+        if status == UCS_ERR_BUSY:
+            return False
+        assert_ucs_status(status)
+        return True
+
+    def ep_create(self, str ip_address, port):
+        cdef ucp_ep_params_t params
+        if c_util_get_ucp_ep_params(&params, ip_address.encode(), port):
+            raise MemoryError("Failed allocation of ucp_ep_params_t")
+
+        cdef ucp_ep_h ucp_ep
+        cdef ucs_status_t status = ucp_ep_create(self._handle, &params, &ucp_ep)
+        c_util_get_ucp_ep_params_free(&params)
+        assert_ucs_status(status)
+        return ucx_ep_create(ucp_ep)
+
+    def progress(self):
+        while ucp_worker_progress(self._handle) != 0:
+            pass
+
+    @property
+    def handle(self):
+        return int(<uintptr_t><void*>self._handle)
+
+    def request_cancel(self, ucp_request_as_int):
+        cdef ucp_request *req = <ucp_request*><uintptr_t>ucp_request_as_int
+
+        # Notice, `ucp_request_cancel()` calls the send/recv callback function,
+        # which will handle the request cleanup.
+        ucp_request_cancel(self._handle, req)
+
+    def tag_recv(self, buffer, size_t nbytes, ucp_tag_t tag, cb_func, cb_args):
+        cdef void *data = <void*><uintptr_t>(get_buffer_data(
+            buffer, check_writable=True)
+        )
+        cdef ucp_tag_recv_callback_t _tag_recv_cb = (
+            <ucp_tag_recv_callback_t>_ucx_tag_recv_callback
+        )
+        cdef ucs_status_ptr_t status = ucp_tag_recv_nb(self._handle,
+                                                       data,
+                                                       nbytes,
+                                                       ucp_dt_make_contig(1),
+                                                       tag,
+                                                       -1,
+                                                       _tag_recv_cb)
+        return handle_comm_result(
+            status, {"cb_func": cb_func, "cb_args": cb_args}, expected_receive=nbytes
+        )
+
+
+cdef class UCXEndpoint:
+    """Python representation of `ucp_ep_h`
+    Please use `ucx_ep_create()` to contruct an instance of this class
+    """
+    cdef:
+        ucp_ep_h _handle
+        bint initialized
+
+    def close(self, UCXWorker worker):
+        cdef ucs_status_ptr_t status
+        if self.initialized:
```

Comment:
I see `self.initialized` is set in `ucx_ep_create`, but what if it isn't called? Is it guaranteed that the default for `bint` is `False`?

Reply:
Yes that is guaranteed, but to make it more readable I have added code that initialize them explicitly.